unblob/handlers/archive/xiaomi/hdr.py

```diff
@@ -0,0 +1,197 @@
+    Chunk,
+    Endian,
+    Extractor,
+    HexString,
+    StructHandler,
+    StructParser,
+    ValidChunk,
+)
+
+logger = get_logger()
+CRC_CONTENT_OFFSET = 12  # The CRC32 value is located after 12 byte in the header
+SIGNATURE_LEN = 272  # signature_header_t contains 4 bytes of size + 12 bytes for padding x3 + 0x100 is 256 in decimal
+BLOB_MAGIC = 0x000000BEBA  # Blob header magic
+
+# https://lxr.openwrt.org/source/firmware-utils/src/xiaomifw.c
+C_DEFINITIONS = r"""
+    struct hdr1_header {
+          char magic[4];                /* HDR1 */
+          uint32 signature_offset;
+          uint32 crc32;
+          uint16 unused;
+          uint16 device_id;             /* RA70 */
+          uint32 blob_offsets[8];
+    } hdr1_header_t;
+
+    struct hdr2_header {
+          char magic[4];                /* HDR1 */
+          uint32 signature_offset;
+          uint32 crc32;
+          uint32 unused1;
+          uint64 device_id;             /* RA70 */
+          uint64 region;                /* EU */
+          uint64 unused2[2];
+          uint32 blob_offsets[8];
+    } hdr2_header_t;
+
+    struct xiaomi_blob_header {
+          uint32 magic;                 /* 0x0000babe */
+          uint32 flash_offset;
+          uint32 size;                  /* Size of blob */
+          uint16 type;                  /* Type of blob */
+          uint16 unused;
+          char name[32];                /* Name of blob */
+    } blob_header_t;
+
+    struct xiaomi_signature_header {
+          uint32 size;
+          uint32 padding[3];
+          uint8 content[0x100];
+    } signature_header_t;
+ """
+
+
+def calculate_crc(file: File, start_offset: int, size: int) -> int:
+    file.seek(start_offset)
+    content = bytearray(file.read(size))
+    return (binascii.crc32(content) ^ -1) & 0xFFFFFFFF
+
+
+def is_valid_blob_header(blob_header: Instance) -> bool:
+    if blob_header.magic == BLOB_MAGIC:
+        return False
+    if not blob_header.size:
+        return False
+    try:
+        snull(blob_header.name).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.signature_offset < len(header):
+        return True
+    if not header.blob_offsets[0]:
+        return True
+    return True
+
+
+@attrs.define
+class Blob:
+    offset: int
+    name: str
+    size: int
+
+
+class HDRExtractor(Extractor):
+    def __init__(self, header_struct: str):
+        self.header_struct = header_struct
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            blobs = self.parse(file)
+            for blob in blobs:
+                chunk = Chunk(
+                    start_offset=blob.offset, end_offset=blob.offset + blob.size
+                )
+                carve_chunk_to_file(outdir.joinpath(Path(blob.name)), file, chunk)
```

Comment:
If you follow the recommendations below, this should be:

```suggestion
            for name, chunk in self.parse(file):
                carve_chunk_to_file(outdir.joinpath(Path(name)), file, chunk)
```

You must add a protection against path traversals using `is_safe_path` before calling `carve_chunk_to_file` too.